modules/xcm-bridge-hub/src/lib.rs

```diff
@@ -0,0 +1,1727 @@
+		/// bridges with remote destinations.
+		// TODO: there's only one impl of `EnsureOrigin<Success = MultiLocation>` -
+		// `EnsureXcmOrigin`, but it doesn't do what we need. Is there some other way to check
+		// `Origin` and get matching `MultiLocation`???
+		type AllowedOpenBridgeOrigin: EnsureOrigin<
+			<Self as SystemConfig>::RuntimeOrigin,
+			Success = MultiLocation,
+		>;
+		/// A converter between a multi-location and a sovereign account.
+		type BridgeOriginAccountIdConverter: ConvertLocation<Self::AccountId>;
+
+		/// Amount of this chain native tokens that is reserved on the sibling parachain account
+		/// when bridge open request is registered.
+		type BridgeReserve: Get<BalanceOf<ThisChainOf<Self, I>>>;
+		/// Currency used to paid for bridge registration.
+		type NativeCurrency: ReservableCurrency<Self::AccountId>;
+
+		/// Delay before bridge is closed.
+		type BridgeCloseDelay: Get<Self::BlockNumber>;
+
+		/// Bridge limits that the regular bridge must not exceed.
+		type BridgeLimits: Get<BridgeLimits>;
+		/// The penalty (in chain native tokens) that is paid from the bridge reserve to the bridge
+		/// misbehavior reporter.
+		type Penalty: Get<BalanceOf<ThisChainOf<Self, I>>>;
+	}
+
+	/// An alias for the bridge metadata.
+	pub type BridgeOf<T, I> = Bridge<ThisChainOf<T, I>>;
+	/// An alias for the bridge state.
+	pub type BridgeStateOf<T, I> = BridgeState<BlockNumberOf<ThisChainOf<T, I>>>;
+	/// An alias for the this chain.
+	pub type ThisChainOf<T, I> =
+		pallet_bridge_messages::ThisChainOf<T, <T as Config<I>>::BridgeMessagesPalletInstance>;
+	/// An alias for the associated lanes manager.
+	pub type LanesManagerOf<T, I> =
+		pallet_bridge_messages::LanesManager<T, <T as Config<I>>::BridgeMessagesPalletInstance>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		T: frame_system::Config<
+			AccountId = AccountIdOf<ThisChainOf<T, I>>,
+			BlockNumber = BlockNumberOf<ThisChainOf<T, I>>,
+		>,
+		T::NativeCurrency: Currency<T::AccountId, Balance = BalanceOf<ThisChainOf<T, I>>>,
+	{
+		/// Open a bridge between two locations.
+		///
+		/// The caller must be within the `T::AllowedOpenBridgeOrigin` filter (presumably: a sibling
+		/// parachain or a parent relay chain). The `bridge_destination_relative_location` must be a
+		/// destination within the consensus of the `T::BridgedNetworkId` network.
+		///
+		/// The `BridgeReserve` amount is reserved on the caller account. This reserve
+		/// is unreserved after bridge is closed.
+		///
+		/// The states after this call: bridge is `Opened`, outbound lane is `Opened`, inbound lane
+		/// is `Opened`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::zero())] // TODO: https://github.com/paritytech/parity-bridges-common/issues/1760 - weights
+		pub fn open_bridge(
+			origin: OriginFor<T>,
+			bridge_destination_relative_location: MultiLocation, // TODO: versioned?
```

Comment:
Another couple of dumb XCM questions:
1) versioned or not - those calls will be called by other chains, so I guess they should be versioned, right? But what are the best practices for that?
2) there's a code where `MultiLocation`s and `InteriorMultiLocation`s are `Box`ed, but e.g. all methods of `MultiLocation` that are accepting `MultiLocation`s/`InteriorMultiLocation`s (e.g. `MultiLocation::reanchor` by value are accepting regular structs, not the `Box`ed versions. I'm tending to use `Box` across my code (because `size_of(InteriorMultiLocation)` is close to 1.3Kb?) but thought if maybe there are some best practices too?

Reply:
I had some discussion about xcm versioning in XCM [channel](https://matrix.to/#/!nYxxyvMNMUaniRIokh:parity.io/$n73az2Zpeco6EfhYNUB12I1DyigA6TrjJAuFO18Hs9I?via=parity.io&via=web3.foundation),
so iiuc when it goes about "public interfaces" and storage then `Versioned*` stuff should be used,
and then internally converted to the `xcm::latest::*` - this is how it is used "everywhere"